### ISSUE TITLE
Tensorflow 1.3: tf.constant with dtype=[float32, float64, float16] may have inconsistent behavior.

### ISSUE THREAD

**Issue by Ouwen** [tensorflow/tensorflow#13827]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Ubuntu 16.04 with docker running `gcr.io/tensorflow/tensorflow:latest` 

- **TensorFlow installed from (source or binary)**: NA
- **TensorFlow version (use command below)**: ('v1.3.0-rc2-20-g0787eee', '1.3.0')
- **Python version**: 2.7
- **Bazel version (if compiling from source)**: NA
- **CUDA/cuDNN version**: NA
- **GPU model and memory**: NA
- **Exact command to reproduce**:

```
// works
test = numpy.array([1,2,3,4,5,6, None], dtype=numpy.float32)
sess = tf.Session()
print(sess.run(tf.constant(test, dtype=tf.float32)))
```

```
// works
sess = tf.Session()
print(sess.run(tf.constant([1, 2, 3, 4, 5, 6, None], dtype=tf.float16)))
```
```
// returns error
sess = tf.Session()
print(sess.run(tf.constant([1, 2, 3, 4, 5, 6, None], dtype=tf.float32)))
// TypeError: Expected float32, got None of type '_Message' instead.
```

### Describe the problem
A tensorflow constant with None in array with dtype float32, float64 seem to throw an error. However, if they are first wrapped by a numpy array, none is accepted and turned into NaN. This behavior seems inconsistent.

**Comment by drpngx** [comment]
/CC @MarkDaoust 

That's unfortunate. Would you like to contribute a fix?

**Comment by Ouwen** [comment]
I'd love to take a stab at contributing. From a brief glance i'm wondering if the `convert_to_tensor` function may be culprit. Glad to take any pointers as well.

**Comment by Ouwen** [comment]
So here is what I've found after looking through the code. This error is propagated from tensorflow version `0.5.0` to `1.4.0-rc0`

`tf.float16` `None` values pass without any problems because they are not checked in [`_TF_TO_IS_OK`](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/python/framework/tensor_util.py#L273-L291). `None` values are thus not stopped by the [`_AssertCompatible`](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/python/framework/tensor_util.py#L376) call, where they continue onward to be transformed into `nan` values by [`np.array`](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/python/framework/tensor_util.py#L377)

For `float32`, and  `float64` types, `None` values are caught by the [`_AssertCompatible`](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/python/framework/tensor_util.py#L294-L303), and since they are not an instance of [`compat.real_types`](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/python/framework/tensor_util.py#L243) `(_numbers.Real, _np.integer, _np.floating)` a `TypeError` is raised.

So while the following produces an error: 
```
sess = tf.Session()
print(sess.run(tf.constant([1, 2, 3, 4, 5, 6, None], dtype=tf.float32)))
// TypeError: Expected float32, got None of type '_Message' instead.
```

Using explicit `nan` will not cause an error
```
sess = tf.Session()
print(sess.run(tf.constant([1, 2, 3, 4, 5, 6, float('nan')], dtype=tf.float32)))
```

Since the `convert_to_tensor` abstraction states the [following are equivalent](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/python/framework/ops.py#L575-L578)
```
value_1 = my_func(tf.constant([[1.0, 2.0], [3.0, 4.0]]))
value_2 = my_func([[1.0, 2.0], [3.0, 4.0]])
value_3 = my_func(np.array([[1.0, 2.0], [3.0, 4.0]], dtype=np.float32))
```
I believe that the below should be equivalent as well
```
value_a = my_func(tf.constant([[1.0, 2.0], [3.0, None]], dtype=tf.float32))
value_b = my_func(np.array([[1.0, 2.0], [3.0, None]], dtype=np.float32))
```

Given what has been found, I would like to
1. Update [`_FliterFloat`](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/python/framework/tensor_util.py#L240-L243) to allow for `None` values.

2. Add `tf.float16` into the  [`_TF_TO_IS_OK`](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/python/framework/tensor_util.py#L273-L291) dictionary with the `_FliterFloat` function

3. Add a test into [`ops_test.py`](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/python/framework/ops_test.py) named `testConvertToTensorFloatNoneValue`

Please let me know if this is a good plan assuming nothing breaks.

**Comment by MarkDaoust** [comment]
Thanks Ouwen! This is the best kind of bug report.

I'm not an expert on this part of TensorFlow, but this all looks pretty reasonable. Please send a PR after you fix that missing `(self)`.

**Comment by Ouwen** [comment]
Appreciate it @MarkDaoust, here is the PR https://github.com/tensorflow/tensorflow/pull/13834

**Comment by MarkDaoust** [comment]
This is fixed by #13834.
